### PR TITLE
Update comments / docs for `Spanned`

### PR DESCRIPTION
**Pull request by alamb**
I took a pass over the code in https://github.com/apache/datafusion-sqlparser-rs/pull/1435 and found some places
I thought adding some more documentation would be helpful.

Changes
1. Updates rustdocs, and adds a bunch of examples
2. Adds links to https://github.com/apache/datafusion-sqlparser-rs/issues/1548

Follow on work (tickets)
- [x] https://github.com/apache/datafusion-sqlparser-rs/pull/1562
- [ ] https://github.com/apache/datafusion-sqlparser-rs/issues/1563

### CODE REVIEW

File: src/ast/helpers/attached_token.rs [hunk 1]
```diff
@@ -80,3 +128,9 @@ impl From<TokenWithLocation> for AttachedToken {
         AttachedToken(value)
     }
 }
+
+impl From<AttachedToken> for TokenWithLocation {
```
Comment (alamb) [inline src/ast/helpers/attached_token.rs:132]: this seemed to be an obvious gap so I just added it to avoid a papercut

File: src/ast/spans.rs [hunk 2]
```diff
@@ -21,21 +21,51 @@ use super::{
...
+///
+/// 1. keywords or other tokens
+/// 2. span information entirely, in which case they return [`Span::empty()`].
+///
+/// Note Some impl blocks (rendered below) are annotated with which nodes are
+/// missing spans. See [this ticket] for additional information and status.
+///
+/// [this ticket]: https://github.com/apache/datafusion-sqlparser-rs/issues/1548
+///
+/// # Example
```
Comment (alamb) [inline src/ast/spans.rs:41]: I felt an example will help people understand how to use the Spans

File: src/lib.rs [hunk 3]
```diff
@@ -61,13 +64,67 @@
...
 //! ```
-//!
 //! [sqlparser crates.io page]: https://crates.io/crates/sqlparser
 //! [`Parser::parse_sql`]: crate::parser::Parser::parse_sql
 //! [`Parser::new`]: crate::parser::Parser::new
 //! [`AST`]: crate::ast
 //! [`ast`]: crate::ast
 //! [`Dialect`]: crate::dialect::Dialect
+//!
+//! # Source Spans
```
Comment (alamb) [inline src/lib.rs:74]: This content is largely migrated from `docs/source_spans.md`

File: src/tokenizer.rs [hunk 4]
```diff
@@ -422,13 +422,35 @@ impl fmt::Display for Whitespace {
...
+/// ```
+/// # use sqlparser::tokenizer::Location;
+/// let location = Location::from((1, 1));
+/// ```
 #[derive(Eq, PartialEq, Hash, Clone, Copy, Ord, PartialOrd)]
 #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
 #[cfg_attr(feature = "visitor", derive(Visit, VisitMut))]
 pub struct Location {
-    /// Line number, starting from 1
+    /// Line number, starting from 1.
```
Comment (alamb) [inline src/tokenizer.rs:447]: `///` means these comments show up in docs

File: src/tokenizer.rs [hunk 5]
```diff
@@ -448,10 +470,25 @@ impl fmt::Debug for Location {
...
+
+    /// Create a new `Location` for a given line and column
+    pub fn new(line: u64, column: u64) -> Self {
         Self { line, column }
     }
 
+    /// Create a new location for a given line and column
+    ///
+    /// Alias for [`Self::new`]
+    // TODO: remove / deprecate in favor of` `new` for consistency?
```
Comment (alamb) [inline src/tokenizer.rs:486]: I found the `Location::of` API kind of confusing -- I would expect that creating a new Rust object would be done via `::new()` so I added a `new()`

I am thinking perhaps we can deprecate `Location::of` maybe to help downstream users

File: src/ast/spans.rs [hunk 6]
```diff
@@ -21,21 +21,51 @@ use super::{
 
 /// Given an iterator of spans, return the [Span::union] of all spans.
 fn union_spans<I: Iterator<Item = Span>>(iter: I) -> Span {
-    iter.reduce(|acc, item| acc.union(&item))
-        .unwrap_or(Span::empty())
+    Span::union_iter(iter)
```
Comment (alamb) [inline src/ast/spans.rs:24]: this was useful enough I felt making it a method in `Span` with more documentation and examples would make it easier to find

File: src/tokenizer.rs [hunk 7]
```diff
@@ -512,16 +567,61 @@ impl Span {
...
+    ///     Span::new(Location::new(1, 1), Location::new(2, 5)),
+    ///     Span::new(Location::new(2, 3), Location::new(3, 7)),
+    ///     Span::new(Location::new(3, 1), Location::new(4, 2)),
+    /// ];
+    /// // line 1, column 1 -> line 4, column 2
+    /// assert_eq!(
+    ///   Span::union_iter(spans),
+    ///   Span::new(Location::new(1, 1), Location::new(4, 2))
+    /// );
+    pub fn union_iter<I: IntoIterator<Item = Span>>(iter: I) -> Span {
```
Comment (alamb) [inline src/tokenizer.rs:596]: This was a private function and I thought it was useful enough it should be its own API and have an example, so I pulled it into a method

**Review by lovasoa — APPROVED** [review]

File: src/ast/mod.rs [hunk 8]
```diff
@@ -596,9 +596,21 @@ pub enum CeilFloorKind {
 
 /// An SQL expression of any type.
 ///
+/// # Semantics / Type Checking
+///
 /// The parser does not distinguish between expressions of different types
-/// (e.g. boolean vs string), so the caller must handle expressions of
-/// inappropriate type, like `WHERE 1` or `SELECT 1=1`, as necessary.
+/// (e.g. boolean vs string). The the caller is responsible for detecting and
```
Comment (iffyio) [inline src/ast/mod.rs:602]:
```suggestion
/// (e.g. boolean vs string). The caller is responsible for detecting and
```

File: src/lib.rs [hunk 9]
```diff
@@ -61,13 +64,67 @@
...
+//! better error messages. See [issue #1548] for more information and the
+//! [`Spanned`] traite to access the spans.
+//!
+//! [issue #1548]: https://github.com/apache/datafusion-sqlparser-rs/issues/1548
+//! [`Spanned`]: ast::Spanned
+//!
+//! ## Migration Guide
+//!
+//! For the next few releases, we will be incrementally adding source spans to the
+//! AST odes, trying to minimize the impact on existing users. Some breaking
```
Comment (iffyio) [inline src/lib.rs:87]:
```suggestion
//! AST nodes, trying to minimize the impact on existing users. Some breaking
```

File: src/lib.rs [hunk 10]
```diff
@@ -61,13 +64,67 @@
...
 //! [`AST`]: crate::ast
 //! [`ast`]: crate::ast
 //! [`Dialect`]: crate::dialect::Dialect
+//!
+//! # Source Spans
+//!
+//! Starting with version  `0.53.0` sqlparser introduced source spans to the
+//! AST. This feature provides source information for syntax errors enables
+//! better error messages. See [issue #1548] for more information and the
+//! [`Spanned`] traite to access the spans.
```
Comment (iffyio) [inline src/lib.rs:79]:
```suggestion
//! [`Spanned`] trait to access the spans.
```

File: src/lib.rs [hunk 11]
```diff
@@ -61,13 +64,67 @@
...
 //! [`Parser::parse_sql`]: crate::parser::Parser::parse_sql
 //! [`Parser::new`]: crate::parser::Parser::new
 //! [`AST`]: crate::ast
 //! [`ast`]: crate::ast
 //! [`Dialect`]: crate::dialect::Dialect
+//!
+//! # Source Spans
+//!
+//! Starting with version  `0.53.0` sqlparser introduced source spans to the
+//! AST. This feature provides source information for syntax errors enables
```
Comment (iffyio) [inline src/lib.rs:77]:
```suggestion
//! AST. This feature provides source information for syntax errors, enabling
```
not sure if the edit does justice to the intended sentence, the 'enables' part seemed to read off was why

File: src/tokenizer.rs (hunk 5 above)
Comment (iffyio) [reply src/tokenizer.rs:486]: That sounds reasonable to me!

**Comment by alamb** [comment]
Thanks @iffyio  -- I'll plan to leave this open for a few more days in case anyone else wants a chance to comment

**Comment by alamb** [comment]
🚀 
Let's get this in

**Comment by alamb** [comment]
Thanks again @iffyio  and @lovasoa

**Pull request merged**